### PR TITLE
Turn down `gp_resource_group_cpu_limit` may cause the database to fail to start

### PR DESCRIPTION
**Pull request by fairyfar**
Turn down `gp_resource_group_cpu_limit` may cause the database to fail to start

# 1. Description

Turn down `gp_resource_group_cpu_limit` may cause the database to fail to start

Firstly, set `gp_resource_manager=group` and the `cpu_max_percent` for any resource group is set to non '-1'.
When we turn down GUC `gp_resource_group_cpu_limit`, then restart GPDB, which may cause GPDB to fail to start.

Root Cause: In cgroup v1, the child cgroup's `cpu.cfs_period_us` must not be larger than the value of the parent cgroup.
When we turn down `gp_resource_group_cpu_limit`, GPDB restarts to set a new `cpu.cfs_period_us` for GPDB's root cgroup,
but the new value may be less than the existing child cgroup's `cpu.cfs_period_us`.

Solution: Reset `cpu.cfs_period_us` to '-1' for all child cgroups before setting GPDB's root cgroup `cpu.cfs_period_us`.

Co-authored-by: @maliangzhu

# 2. Test case

## System infos

```bash
# OS version:
[root@bogon ~]# cat /etc/os-release
NAME="Red Hat Enterprise Linux Server"
VERSION="7.6 (Maipo)"
ID="rhel"

# GPDB version:
[yz@bogon gpdb]$ psql postgres
psql (12.12)
Type "help" for help.

postgres=# select version();
                                                                                                                   version

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.1+dev.1237.gb741d59cfb build dev) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2), 64-bit compiled on Sep 28 2023 17:50:43 (with assert checking) Bhuvnesh C.
(1 row)

# There are 4 CPU cores on the system.
[root@bogon ~]# cat /proc/cpuinfo | grep processor | wc -l
4
```

## Enable resource group manager & restart MPP

```bash
gpconfig -c gp_resource_manager -v group
gpstop -r -a
```

## Configure group

```bash
postgres=# create resource group rg with (cpu_max_percent=80);
CREATE RESOURCE GROUP
postgres=# select * from gp_toolkit.gp_resgroup_config;
 groupid |   groupname   | concurrency | cpu_max_percent | cpu_weight | cpuset | memory_limit | min_cost | io_limit
---------+---------------+-------------+-----------------+------------+--------+--------------+----------+----------
    6437 | default_group | 20          | 20              | 100        | -1     | -1           | 500      | -1
    6438 | admin_group   | 10          | 10              | 100        | -1     | -1           | 500      | -1
    6441 | system_group  | 0           | 10              | 100        | -1     | -1           | 500      | -1
   18639 | rg            | 20          | 80              | 100        | -1     | -1           | 0        | -1
(4 rows)
```

## Check cgroup values

```bash
[root@bogon ~]#cd /sys/fs/cgroup/cpu
[root@bogon cpu]# cat cpu.cfs_quota_us
-1
[root@bogon cpu]# cat gpdb/cpu.cfs_quota_us
360000
[root@bogon cpu]# cat gpdb/18639/cpu.cfs_quota_us
288000
```

## Changed gp_resource_group_cpu_limit and restart MPP

```bash
[yz@bogon gpdb]$ gpconfig -c gp_resource_group_cpu_limit -v 0.6

[yz@bogon gpdb]$ gpstop -a
......
[yz@bogon gpdb]$ gpstart -a
20231211:15:28:53:009706 gpstart:bogon:yz-[INFO]:-Starting gpstart with args: -a
20231211:15:28:53:009706 gpstart:bogon:yz-[INFO]:-Gathering information and validating the environment...
20231211:15:28:53:009706 gpstart:bogon:yz-[INFO]:-Greenplum Binary Version: 'postgres (Greenplum Database) 7.0.0-beta.1+dev.1237.gb741d59cfb build dev'
20231211:15:28:53:009706 gpstart:bogon:yz-[INFO]:-Greenplum Catalog Version: '302307241'
20231211:15:28:53:009706 gpstart:bogon:yz-[INFO]:-Starting Coordinator instance in admin mode
20231211:15:28:53:009706 gpstart:bogon:yz-[INFO]:-CoordinatorStart pg_ctl cmd is env GPSESSID=0000000000 GPERA=None $GPHOME/bin/pg_ctl -D /home/yz/gpdb/db/coordinator/gpseg-1 -l /home/yz/gpdb/db/coordinator/gpseg-1/log/startup.log -w -t 600 -o " -c gp_role=utility " start
20231211:15:28:53:009706 gpstart:bogon:yz-[CRITICAL]:-Failed to start Coordinator instance in admin mode
20231211:15:28:53:009706 gpstart:bogon:yz-[CRITICAL]:-Error occurred: non-zero rc: 1
 Command was: 'env GPSESSID=0000000000 GPERA=None $GPHOME/bin/pg_ctl -D /home/yz/gpdb/db/coordinator/gpseg-1 -l /home/yz/gpdb/db/coordinator/gpseg-1/log/startup.log -w -t 600 -o " -c gp_role=utility " start'
rc=1, stdout='waiting for server to start.... stopped waiting
', stderr='pg_ctl: could not start server
Examine the log output.
'

[yz@bogon gpdb]$ vim /home/yz/gpdb/db/coordinator/gpseg-1/log/startup.log
......
2023-12-11 15:28:53.690474 CST,,,p9716,th228710528,,,,0,,,seg-1,,,,,"FATAL","XX000","can't write data to file '/sys/fs/cgroup/cpu/gpdb/cpu.cfs_quota_us': 无效的参数",,,,,,,,"writeData","cgroup.c",341,1    0xeac7f3 postgres errstart (elog.c:484)
2    0xf0b3f1 postgres writeData (cgroup.c:341)
3    0xf0b61b postgres writeInt64 (cgroup.c:386)
4    0xf0c6d1 postgres <symbol not found> (cgroup-ops-linux-v1.c:445)
5    0xf0c9cc postgres <symbol not found> (cgroup-ops-linux-v1.c:615)
6    0xf02536 postgres initCgroup (resgroup.c:424)
7    0xc3b13e postgres PostmasterMain (postmaster.c:1538)
8    0xaf6d7a postgres <symbol not found> (main.c:259)
9    0x7f5c0b196555 libc.so.6 __libc_start_main + 0xf5
10   0x7160cf postgres <symbol not found> + 0x7160cf

2023-12-11 15:28:53.830224 CST,,,p9716,th228710528,,,,0,,,seg-1,,,,,"LOG","00000","database system is shut down",,,,,,,,"UnlinkLockFiles","miscinit.c",919,
```

# Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by SmartKeyerror** [comment]
This change looks good to me, will approve it after I run some tests and pass all test cases of the pipeline.

File: src/backend/utils/resgroup/cgroup-ops-linux-v1.c [hunk 1]
```diff
@@ -397,6 +399,54 @@ check_component_hierarchy_v1()
...
+
+	while ((de = readdir(root_dir)) != NULL)
+	{
+		if (strcmp(de->d_name, ".") == 0 || strcmp(de->d_name, "..") == 0)
+			continue;
+
+		/* sub directory */
+		snprintf(dest_path, sizeof(dest_path), "%s%s", root_path, de->d_name);
+
+		struct stat statbuf;
```
Comment (SmartKeyerror) [inline src/backend/utils/resgroup/cgroup-ops-linux-v1.c:430]: Please do not define a new variable in this place, move it to the head of the function.

Comment (fairyfar) [reply src/backend/utils/resgroup/cgroup-ops-linux-v1.c]: I have modified it.

**Comment by SmartKeyerror** [comment]
Hi @fairyfar , thanks for your contribution! 

Please merge the current main branch to update the concourse pipeline config file, then I can run dev pipeline after that. I intend to do this but it seems that you did not allow the maintainer to modify this PR, so you need to merge the latest code, thanks a lot.

**Comment by fairyfar** [comment]
> Hi @fairyfar , thanks for your contribution!
> 
> Please merge the current main branch to update the concourse pipeline config file, then I can run dev pipeline after that. I intend to do this but it seems that you did not allow the maintainer to modify this PR, so you need to merge the latest code, thanks a lot.

Nailed it.

**Review by SmartKeyerror — APPROVED** [review]

**Pull request merged**